Reverse proxy dynamic subpath

fabio-torchetti opened the pull request:
## What is it?

This is a PR to allow dynamic changing of the subpath the server is running behind. The idea is to allow the server to adapt to being behind a proxy that changes the root path.

## How?

The main `index.html` page is rendered on the fly taking into consideration the path the page is being served from. This can be determined in two ways:

  1 - A `X-External-Path` header is passed along with the request
  2 - A `base` is specified in the application configuration

  The api base and `socket.io` path is calculated based on the document `baseURI` is set in the `index.html` base tag.

## Why?

I am planning to have ZWave To MQTT run behind the `ingress` of Home Assistant. I really didn't like the option to expose it directly under a new domain. So, in the spirit of the Open Source _"If you don't like it, fix it!"_ this PR came to be.

This related to #317

fabio-torchetti commented:
## Working using Ingress

![Screen Shot 2020-06-01 at 07 36 36](https://user-images.githubusercontent.com/53441171/83409955-0d07c600-a3db-11ea-9973-77823ac6cb8b.png)

robertsLando commented:
I would like that @chrisns reviews this too before the merge

robertsLando commented on README.md:
```diff
@@ -168,6 +168,11 @@ kubectl apply -k https://raw.githubusercontent.com/openzwave/zwave2mqtt/master/k
 
 4. Open the browser <http://localhost:8091>
 
+### Reverse Proxy Setup
```
Please add this to the TOC

fabio-torchetti replied:
Added to the TOC

chrisns commented on src/apis/ConfigApis.js:
```diff
@@ -1,20 +1,24 @@
 }
 
+axios.defaults.socketUrl = getBasePath()
+axios.defaults.baseURL = axios.defaults.socketUrl + '/api'
```
since you've not concatinated strings like that elsewhere
```suggestion
axios.defaults.baseURL = `${axios.defaults.socketUrl}/api`
```

chrisns commented on src/apis/ConfigApis.js:
```diff
@@ -1,20 +1,24 @@
 import { loadProgressBar } from 'axios-progress-bar'
 
-if (process.env.NODE_ENV === 'development') {
-  axios.defaults.socketUrl = location.protocol + '//' + location.hostname + ':' + process.env.PORT
```
i don't know how this bit of code works or really what it was meant for, have you checked if it breaks the `npm run dev` which uses `webpack-dev-server` ?

fabio-torchetti replied:
I have tested both dev and prod builds. There is still an issue for the dev version: it seems it can't correctly find the zwave logo. All the rest works as expected.

fabio-torchetti replied:
The code - I believe - allowed the dev seerver to correctly point to the websocket path. Right now we are using the namespace + base path combination, so it's no longer needed.

robertsLando replied:
IMO the problem is here

fabio-torchetti replied:
@robertsLando The requests to `/socket.io` are proxied by the webpack webserver to localhost:8091 -- I tested it yesterday, and as long as you have the services running on that port it should work. Let me check what's going on and fix the lint errors.

robertsLando replied:
Did you try to run `npm run dev` and `node start` and go to `localhost:8092`? It doesn't work to me

fabio-torchetti replied:
@robertsLando Try now! Forgot to add `/api` and `/health` to the proxied endpoints. See [this part](https://github.com/OpenZWave/Zwave2Mqtt/pull/540/files#diff-c8c5ee8e781f831e4927a392fa3565e6R13) of the configuration.

  This will have the webpack webserver proxy requests done on port 8092 for `/api`, `/socket.io` and `/health` forwarded to `localhost:8091`. I prefer this solution so that you don't have to handle _development_ cases in production code.

  Give this fixed version a shot and let me know what you think.

  Saluti da un Italiano in Iowa! :)

robertsLando replied:
Just tried now and everything is working as expected! Both `webpack-dev` and `health` endpoints, love your solution, I admit I'm not a webpack expert and I dind't know about that, thanks for this :) I think everything is ready for merge now

chrisns commented on test/lib/renderIndex.test.js:
```diff
@@ -0,0 +1,96 @@
+const chai = require('chai')
```
❤️ 😍 ❤️  first frontend test! whoop!

chrisns commented on views/index.ejs:
```diff
@@ -0,0 +1,40 @@
+    <meta name="viewport" content="width=device-width,initial-scale=1.0">
+    <title><%= config.title %></title>
+    <base href="<%= config.base %>" />
+    <!--Favicons-->
```
the favicons are lower, we don't really need a html comment though for it

fabio-torchetti replied:
Removed -- but I promise I didn't add the comment :)

chrisns commented on README.md:
```diff
@@ -413,6 +418,7 @@ Zwave2Mqtt to resend the cached values when Home Assistant restarts.
 
 Zwave2Mqtt try to do its best to guess how to map devices from Zwave to HASS. At the moment it try to guess the device to generate based on zwave values command classes, index and units of the value. When the discovered device doesn't fit your needs you can you can set custom a `device_class` to values using Gateway value table.
 
+
```
https://github.com/OpenZWave/Zwave2Mqtt/pull/549

fabio-torchetti replied:
Fixed!

fabio-torchetti commented on docs/subpath.md:
````diff
@@ -0,0 +1,53 @@
+You can pass the external path by setting the `X-External-Path` header, for example
+suppose you had the following `nginx` configuration:
+
+```text
````
Fixed

fabio-torchetti commented on docs/subpath.md:
````diff
@@ -0,0 +1,53 @@
+pick up the `X-Ingress-Path;` and map it, something along
+these lines:
+
+```text
````
Fixed

fabio-torchetti commented on app.js:
```diff
@@ -14,9 +14,9 @@ var store = reqlib('config/store.js')
-
 var gw; //the gateway instance
 let io;
+const renderIndex = reqlib('/lib/renderIndex')
```
Fixed -- well spotted -- don't know how it ended up there.

fabio-torchetti commented on docs/subpath.md:
```diff
@@ -0,0 +1,53 @@
+subpaths to serve the pages and services. 
+
+You can use a header to signal where the external path is or you can configure
+the base path. In both cases this are dynamic configurations, so you can deploy
```
Thanks!

fabio-torchetti commented on build/webpack.dev.conf.js:
```diff
@@ -2,6 +2,7 @@
 const utils = require('./utils')
 const webpack = require('webpack')
 const config = require('../config')
+const appConfig = require('../config/webConfig');
```
Fixed.

fabio-torchetti commented:
My only hesitation in merging this is that when running with the fronted dev server the `png` for the ZWave logo can't be found by the frontend. 

  Other than that I am using this branch on my home assistant instance using a [custom addon](https://github.com/fabio-torchetti/addon-zwave2mqtt) based on these changes.

  Once this is merged I will submit a PR for the official addon to use the Ingress friendly version.

robertsLando commented:
My only hesitation in merging this is that when running with the fronted dev server the png for the ZWave logo can't be found by the frontend.

You could add the base path to the environment vars of webpack (like I did with version) and add that to the png path

robertsLando commented:
@chrisns I prefer to wait for #549 to be merged before this and the logo thing to be fixed

chrisns commented:
> @chrisns I prefer to wait for #549 to be merged before this and the logo thing to be fixed

agreed, i'll fix up #549 asap

fabio-torchetti commented:
> @chrisns I prefer to wait for #549 to be merged before this and the logo thing to be fixed

  I think it's a good idea. I have fixed the logo during development testing, added a more complete example of nginx configuration on  a subpath -- it's the configuration I use to test working behind a reverse proxy.

  I have added a development dependency - proxyquire - that allows to test more static classes, like the `webConfig`. I checked the audit and it does not have any outstanding security issues.

  I also added the proxy settings for webpack's webserver, this allows both hot reloading while using a running webserver. This is currently assuming that the server is running locally on port 8091. IE: you should be able to run `npm run dev:server &` then `npm run dev` and end up with a hot reloading web interface on `:8902`.

  Give it a try and let me know what you think.

robertsLando commented:
Grazie Fabio :smile: 

I will try it now give me a moment

robertsLando commented:
@fabio-torchetti When running `npm run dev` command and going to the ui seems all api calls are failing

chrisns commented:
given we've used rewire elsewhere and making that play nice with proxyquire from research is hard had you considered [rewiremock](https://github.com/theKashey/rewiremock)?

chrisns commented:
> given we've used rewire elsewhere and making that play nice with proxyquire from research is hard had you considered [rewiremock](https://github.com/theKashey/rewiremock)?

please ignore ^^ comment

see https://github.com/thlorenz/proxyquire/issues/90#issuecomment-178056535

I'd only introduced rewire really to allow some instrumentation to be able to then write tests without changing the code, its a gross hack and can hopefully be removed at some point in the future.

robertsLando commented:
@fabio-torchetti I saw your latest commit, if you use vscode you should have it auto running on save :)

chrisns commented on .vscode/settings.json:
```diff
@@ -5,5 +5,8 @@
-  "wallaby.startAutomatically": true
+  "wallaby.startAutomatically": true,
+  "[vue]": {
+    "editor.defaultFormatter": "dbaeumer.vscode-eslint"
```
did you have the `numso.prettier-standard-vscode` plugin installed, was it not working? it should cover vue code

fabio-torchetti replied:
Not sure how my `.vscode` folder got committed. My new setup is a workspace with client and server separated - it allows me to use the mocha test explorer with both the node - commonjs - and vue - esm - syntaxes.

  I prefer eslint, but again -- this probably should have not been committed.

chrisns commented on README.md:
```diff
@@ -168,6 +169,11 @@ kubectl apply -k https://raw.githubusercontent.com/openzwave/zwave2mqtt/master/k
+### Reverse Proxy Setup
+
+If you need to setup ZWave To MQTT behind a reverse proxy that needs a _subpath_ to
+work, take a look at [this](docs/subpath.md).
```
semantic links for accessibility are better, imagine a screen reader shouting **THIS** at you when someone is tabbing through :)
```suggestion
work, take a look at [the reverse proxy configuraiton docs](docs/subpath.md).
```

chrisns commented on build/webpack.dev.conf.js:
```diff
@@ -61,8 +55,12 @@ const devWebpackConfig = merge(baseWebpackConfig, {
     new webpack.NoEmitOnErrorsPlugin(),
     // https://github.com/ampedandwired/html-webpack-plugin
     new HtmlWebpackPlugin({
+      title: 'ZWave  To MQTT',
```
extra space snuck in there?
```suggestion
      title: 'ZWave To MQTT',
```

chrisns commented on config/index.js:
```diff
@@ -9,7 +9,14 @@ module.exports = {
-    proxyTable: {},
+    proxyTable: {
+      '/socket.io': {
+        target: 'ws://localhost:8091',
```
does this work if https://github.com/OpenZWave/Zwave2Mqtt/blob/master/config/app.js#L4 is not 8091?

chrisns replied:
also as below, if it can be overridden to listen on **not** `localhost` will this work?

fabio-torchetti replied:
Let me work on this.

fabio-torchetti replied:
Done - documented how to use `SERVER_HOST` - `SERVER_PORT` - `SERVER_SSL` and other environment variables to affect the proxying.

robertsLando replied:
Nice catch @chrisns I completely missed this

chrisns commented on src/App.vue:
```diff
@@ -214,7 +218,9 @@ export default {
     var self = this
 
-    this.socket = io(ConfigApis.getSocketIP())
+    this.socket = this.socket = io('/', {
```
not sure what the double `=` is doing here, mistake or does it have some functionality I don't know about?
```suggestion
    this.socket = io('/', {
```

fabio-torchetti replied:
Nope -- syntactically correct, but a complete oversight!

chrisns commented on test/config/webConfig.test.js:
```diff
@@ -0,0 +1,47 @@
+  })
+
+  describe('Uses defaults if nothing specified', () => {
+    it('uses "/" as the default base', () => {
```
don't need curly braces for all your assertions, removing them and implicitly returning booleans is safer for catching false positives failures

fabio-torchetti replied:
TIL! You will have to talk to me more about this, I have never returned anything from my UT -- and now you made me think about it, and I need to see what I have been missing!

chrisns commented on test/lib/renderIndex.test.js:
```diff
@@ -0,0 +1,102 @@
+        },
+        mockResponse
+      )
+      lastOptions.config.base.should.equal('/test/base/')
```
safer to return the bool
```suggestion
      return lastOptions.config.base.should.equal('/test/base/')
```

robertsLando commented:
OK let's merge this. Thanks again @fabio-torchetti :tada:

chrisns commented:
thanks for bearing with uss @fabio-torchetti that was a change required for what should have been way simpler
